### PR TITLE
adding repoproviders documentation

### PR DESCRIPTION
**Pull request by choldgraf**
This is a documentation update for repository providers, it does these main things:

1. Updates our table of repoprovider specs to include recent additions
2. Adds a new page dedicated to repository providers, and moves a section of the API docs there (with the goal of making this information more discoverable)
3. Adds a section on adding a new repository provider, so it's easier to
   provide guideance on PRs like #969 

What do folks think?

### CODE REVIEW

File: doc/developer/repoproviders.rst [hunk 1]
```diff
@@ -0,0 +1,60 @@
+.. _providers-section:
+
+====================
+Repository Providers
+====================
+
+Repository Providers (or RepoProviders) are
+locations where git repositories are stored (e.g.,
```
Comment (betatim) [inline doc/developer/repoproviders.rst]:
```suggestion
locations where repositories are stored (e.g.,
```
The Zenodo or FigShare providers don't (need to) give us something that is a Git repository (contains a `.git` subdirectory). I think calling the thing we start from a "repository" is a good compromise between using something specific (c.f. "directory" is super vague I think, even though it is strictly true) and trying to move away from the idea that it has to be a git repository.

File: doc/developer/repoproviders.rst [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+| Figshare   | ``figshare``       | ``<figshare-DOI>``                                          |                            |
++------------+--------------------+-------------------------------------------------------------+----------------------------+
+| Git        | ``git``            | ``<url-escaped-url>/<commit-sha>``                          | arbitrary HTTP git repos   |
++------------+--------------------+-------------------------------------------------------------+----------------------------+
+
+Adding a new repository provider
+================================
+
+It is possible to add new repository providers to BinderHub, allowing
+a BinderHub deployment to fetch git repositories from new locations
```
Comment (betatim) [inline doc/developer/repoproviders.rst]:
```suggestion
a BinderHub deployment to fetch repositories from new locations
```

File: doc/developer/repoproviders.rst [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+on the web. Doing so involves defining your own RepoProvider sub-class
+and modifying a set of methods/attributes to interface with the online
+provider to which you are providing access.
+
+In order to extend the supported repository providers,
```
Comment (betatim) [inline doc/developer/repoproviders.rst]: Should we add a step that it is highly likely that your first step would be to add a new content provider to repo2docker?

File: doc/developer/repoproviders.rst (hunk 1 above)
Comment (choldgraf) [reply doc/developer/repoproviders.rst]: good point, totally agree

**Comment by choldgraf** [comment]
accepted your comments and added a note about content providers. I also realized we don't have a similar section for how to add a new content provider in the repo2docker docs, so I opened https://github.com/jupyter/repo2docker/issues/823 too

**Comment by nuest** [comment]
I re-visited a repo provider I've added (see #951) and found that I had to make changes to `binderhub/static/js/index.js` for the example text and `binderhub/templates/index.html` for the dropdown entry. AFAICS that's missing from the steps.

[ The JS file reminds me of the need for a proper API to generate the UI... still very good to document the current state though! ]

**Comment by choldgraf** [comment]
@nuest aren't both of those addressed here: https://github.com/jupyterhub/binderhub/pull/1017/files#diff-cbeb18fbf4548b84e49c010cf02f439dR57 ? Let me know if I'm missing something!

(and I totally agree about the need to do less hard-coding etc)

**Comment by nuest** [comment]
@choldgraf My bad, those files are indeed included in that changeset!

**Comment by choldgraf** [comment]
anything else to do here? If we can get this in then we'll ask #967 and #969 to document themselves and then we can get those merged too

**Comment by choldgraf** [comment]
I thought about it a bit more and decided to add repository provider information to the table, the new version is here:

https://643-89419368-gh.circle-artifacts.com/0/html/developer/repoproviders.html#supported-repoproviders

**Comment by betatim** [comment]
Merging this now. We can keep tweaking it in a new PR.

**Pull request merged**